core/src/com/unciv/logic/multiplayer/chat/ChatWebSocketManager.kt

```diff
@@ -0,0 +1,249 @@
+            protocol = if (protocol.isSecure()) URLProtocol.WSS else URLProtocol.WS
+        }.build()
+
+    /*
+     * Delivery not guaranted & failures are mosly ignored.
+     */
+    fun requestMessageSend(message: Message) {
+        startSocket()
+        Concurrency.run("MultiplayerChatSendMessage") {
+            withTimeoutOrNull(1000) {
+                while (session == null) {
+                    delay(50)
+                    yield()
+                }
+            }
+            session?.runCatching {
+                this.sendSerialized(message)
+            }
+        }
+    }
+
+    fun handleWebSocketThrowables(t: Throwable) {
+        println("ChatError: ${t.message}")
+
+        if (errorReconnectionAttempts == 0)
+            relayGlobalMessage("WebSocket connection closed. Cause: ${t.cause}!")
+
+        if (
+            errorReconnectionAttempts == 0 &&
+            t.message?.contains("401") == true
+        ) {
+            relayGlobalMessage("Authentication issue detected! You have to set a password to use Chat.")
+        }
+
+        restartSocket(true)
+    }
+
+    suspend fun startSession() {
+        try {
+            session?.close()
+            session = client.webSocketSession {
+                url(chatUrl)
+                header(HttpHeaders.Authorization, UncivGame.Current.settings.multiplayer.getAuthHeader())
+            }
+
+            session!!.runCatching {
+                if (isActive) {
+                    if (errorReconnectionAttempts == 0) {
+                        println("ChatLog: Connected to WebSocket.")
+                        relayGlobalMessage("Successfully connected to WebSocket server!")
+                    } else if (errorReconnectionAttempts > 0) {
+                        println("ChatLog: Re-established webSocket connection.")
+                        relayGlobalMessage("Successfully re-established WebSocket connection!")
+                    }
+                }
+
+                val gameIds = ChatStore.getGameIds()
+                    .union(UncivGame.Current.onlineMultiplayer.games.mapNotNull { it.preview?.gameId })
+                this.sendSerialized(Message.Join(gameIds.toList()))
+
+                while (this.isActive) {
+                    val response = receiveDeserialized<Response>()
+                    when (response) {
+                        is Response.Chat -> EventBus.send(
+                            ChatMessageReceived(
+                                response.gameId, response.civName, response.message
+                            )
+                        )
+
+                        is Response.Error -> relayGlobalMessage("Error: ${response.message}", "Server")
+                        is Response.JoinSuccess -> Unit
+                    }
+                    yield()
+                }
+            }
+                .onSuccess { restartSocket() }
+                .onFailure { handleWebSocketThrowables(it) }
+        } catch (e: Exception) {
```

Comment:
The only area not caught is the party above runcatching. Catches should be localized to the minimum viable area

Reply:
What did you mean, I did not get it properly.